spmd/compiler/fusion.py

```diff
@@ -970,3 +935,285 @@ def run_fuse_communication_cat(gm: fx.GraphModule, fusion_length: int) -> None:
+    last_grad_fe_index = _get_last_grad_node_from_fe_group(gi, copy_list)
+
+    last_grad_tensor_node = cast(
+        fx.Node,
+        cast(fx.Node, copy_list[last_grad_fe_index].grad_tensor_node).args[0],
+    )
+
+    jit_inputs = []
+    offset = 0
+    size = 0
+    copy_nodes = []
+
+    buffer_size_needed = sum([item.size for item in copy_list])  # type: ignore
+
+    device = torch.cuda.current_device()
+    gpu = "cuda:" + str(device)
+
+    jit_buffer_node = gm.graph.create_node(
+        OP.CALL_FUNCTION,
+        target=torch.empty,
+        args=(buffer_size_needed,),
+        kwargs={"device": gpu},
+    )
+    jit_inputs.append(jit_buffer_node)
+
+    for fe in copy_list:
+        start = offset
+        stop = offset + fe.size
+        # jump from clone to actual tensor node
+        grad_tensor_clone_node = cast(fx.Node, fe.grad_tensor_node)
+        grad_node = grad_tensor_clone_node.args[0]
+
+        view_node = gm.graph.call_function(
+            torch.ops.aten.view.default,
+            (
+                grad_node,
+                [-1],
+            ),
+        )
+
+        slice_node = gm.graph.call_function(
+            torch.ops.aten.slice.Tensor,
+            (jit_buffer_node, 0, start, stop),
+        )
+        copy_node = gm.graph.call_function(
+            # torch.Tensor.copy_,
+            torch.ops.aten.copy_.default,
+            (slice_node, view_node),
+        )
+        offset += fe.size
+        jit_inputs.extend([view_node, slice_node, copy_node])
+        copy_nodes.append(copy_node)
+
+    assert copy_list[-1].comm_node is not None
+
+    fused_comm_node = copy_list[-1].comm_node  # type: ignore
+
+    fused_comm_node.update_arg(0, [jit_buffer_node])  # type: ignore
+
+    fused_comm_node.users[jit_buffer_node] = ""  # type: ignore
+
+    # Move the fused_comm_node and its args to right after the source node
+    nodes_to_move = jit_inputs + [
+        fused_comm_node.args[1],  # type: ignore
+        fused_comm_node.args[2],  # type: ignore
+        fused_comm_node,
+    ]  # type: ignore
+
+    insert_node = last_grad_tensor_node.next
+    for node in nodes_to_move:
+        insert_node.prepend(node)
+
+    # enforce users count
+    for node in copy_nodes:
+        user = node.args[0]
+        node.users[user] = ""  # type: ignore
+        node_user_len = len(node.users)
+        assert node_user_len, f"failed to update users for node {node.name}"
+
+    return jit_buffer_node
+
+
+def _scatter_results_jit(
+    gi: GraphInfo,
+    gm: fx.GraphModule,
+    scatter_list: List[FusionElement],
+    jit_buffer_node: fx.Node,
+) -> List[fx.Node]:
+    """prepare views against completed buffer, these will replace gradient nodes for
+    graph output to avoid copy back overhead."""
+
+    assert scatter_list[-1].wait_node is not None
+    wait_node = scatter_list[-1].wait_node
+
+    # ensure user
+
+    wait_user = wait_node.args[0]  # type: ignore
+    wait_node.users[wait_user] = ""  # type: ignore
```

Comment:
Why assigning users an empty string instead of popping it out? Any why do we have to modify the `users`?

Reply:
Never mind. You mentioned that in the talk.

Reply:
cool, thanks.  
Just for any future viewing reference - certain graph operations such as copy do not register as having a user, so we have to manually assign it a user or it gets blown away during next recompile.  
The user 'dict' is treated as a set, so only keys count and value = "don't care" ,thus assigning it with "".